core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -147,6 +148,8 @@
 
     private final ConcurrentMapLong<SearchContext> activeContexts = ConcurrentCollections.newConcurrentMapLongWithAggressiveConcurrency();
 
+    private final CopyOnWriteArrayList<Long> releasedContexts = new CopyOnWriteArrayList<>();
```

Comment:
By keeping track of contexts in 2 different data-structures, I think you are potentially introducing race conditions, eg. if a clear scroll action is issued concurrently with an automatic release of the context due to the fact there are no more hits to process.